[CUDA] Asynchronous command submission

jrprice opened the pull request:
This patch makes the CUDA backend deal with command submission asynchronously. This enables commands to be overlapped when using multiple devices, or with multiple command queues targeting the same device (e.g. overlapping compute with data transfers). Event dependencies are now handled using CUDA events instead of blocking on the host.

The `events` category of the Khronos OpenCL conformance test suite now passes for the CUDA backend. I've also added a small section to the CUDA docs that list all the conformance test categories that pass.

There's a small handful of changes and bug fixes outside of the CUDA backend, but I don't think there's anything controversial.

pjaaskel commented on lib/CL/devices/cuda/pocl-cuda.c:
```diff
@@ -557,34 +660,42 @@ pocl_cuda_copy_rect (void *data, const void *__restrict const src_ptr,
-  CUresult result = cuMemcpyDtoH (ptr, (CUdeviceptr) (buf_ptr + offset), size);
-  CUDA_CHECK (result, "cuMemcpyDtoH");
-  return ptr;
+  // TODO: Map instead of copy?
```
Please use old C style comments.

pjaaskel commented on lib/CL/devices/cuda/pocl-cuda.c:
```diff
@@ -557,34 +660,42 @@ pocl_cuda_copy_rect (void *data, const void *__restrict const src_ptr,
-          (CUdeviceptr) (device_start_ptr + offset), host_ptr, size);
-      CUDA_CHECK (result, "cuMemcpyHtoD");
-      free (host_ptr);
+      // TODO: Only copy back if mapped for writing
```
Ditto.

jrprice commented:
Fixed comments - all C-style now in that file.

Also snuck in one more low-hanging conformance test fix.

pjaaskel commented:
There are some failing pthread tests (at least with llvm-trunk):
```
83:runtime/clCreateSubDevices
86:runtime/test_user_event
115:AMD_30_AdvancedConvolution
117:AMD_30_BinomialOption
118:AMD_30_BinomialOptionMultiGPU
119:AMD_30_BitonicSort
120:AMD_30_BlackScholes
121:AMD_30_BlackScholesDP
122:AMD_30_BoxFilter
124:AMD_30_ConcurrentKernel
125:AMD_30_DCT
128:AMD_30_DwtHaar1D
129:AMD_30_FastWalshTransform
130:AMD_30_FloydWarshall
139:AMD_30_LUDecomposition
140:AMD_30_MatrixMulImage
141:AMD_30_MatrixMultiplication
142:AMD_30_MatrixTranspose
144:AMD_30_MonteCarloAsian
145:AMD_30_MonteCarloAsianMultiGPU
147:AMD_30_PrefixSum
149:AMD_30_RadixSort
150:AMD_30_RecursiveGaussian
151:AMD_30_Reduction
153:AMD_30_SimpleConvolution
156:AMD_30_SobelFilter
157:AMD_30_StringSearch
158:AMD_30_Template
160:AMD_30_URNG
```

jrprice commented:
Should be fixed now.

I've also addressed the issues we were talking about #478 by adding a second background thread to handle command completion. This also greatly simplifies a lot of the code, and we no longer need to explicitly check the event wait list. The CUDA backend now behaves a lot more like the other device drivers, relying on the broadcast/notify approach to signal event completion.

vkorhonen commented:
It seems that notifications and broadcasting are doing what they supposed to. Except for notify seems to assume that notifier event is not from this particular cuda device (`assert (event_data->num_ext_events > 0);`). The original idea is that all events are notified, so the device may notify itself. I do not see any events being removed from the wait lists. I might have misunderstood something, because you should have hit that assert in many cases already, assuming my assumption is correct.

In function `pocl_cuda_submit_node` you might want to lock the dep event before checking its status and unlocking it after `event_data->num_ext_events++;` to ensure that increment happens before dep broadcast. That `num_ext_events` should probably be volatile since it is possibly updated by another driver thread.

Submit/finalize-threads are continuously running full speed. `pthread_cond_wait` could be used there to suspend the threads when there is nothing to do. Also wait_event could use condition variable (see pthread implementation).

I am not sure if mem migration works. I assume that `cuMemcpyDtoDAsync` copies data around inside a device and the purpose of the migrate is to transfer mem object from another device/host to this device. So you might want to use `pocl_migrate_mem_objects` which uses this device drivers copy function if src and the dst are in the same global memory or else it uses appropriate drivers APIs to read from src device to host memory and then write to dst device. 
Actually the same applies to all copy commands, it may be that src dev != dst dev. This is quite important to do correctly if someone needs to implement for example cpu<->gpu pipelined application.

jrprice commented:
Thanks for the feedback @vkorhonen.

I've fixed the issue regarding notifications from CUDA devices to themselves in PR #485. That PR also deals with the memory migration issue you raised, although it still won't work if one of the devices is using non-host memory.

> In function pocl_cuda_submit_node you might want to lock the dep event before checking its status and unlocking it after event_data->num_ext_events++; to ensure that increment happens before dep broadcast. That num_ext_events should probably be volatile since it is possibly updated by another driver thread.

IIRC this introduces a deadlock, but actually shouldn't be necessary. The dep event won't be able to broadcast to this node while we're in the process of incrementing `num_ext_events` because it would have to lock this event first. This was really awkward to reason about but I *think* it works.

> Submit/finalize-threads are continuously running full speed. pthread_cond_wait could be used there to suspend the threads when there is nothing to do. Also wait_event could use condition variable (see pthread implementation).

I've made a note of this and will give it a go when I've regained access to some systems where I can test this properly in a multi-device set-up.

pjaaskel commented:
Is this done?
